judge/runpipe.cc

```diff
@@ -1,600 +1,858 @@
+               write_end, j, read_end);
+        process.stdout = write_end;
+        other.stdin = read_end;
+      }
+    }
+  }
+
+  // Create the epoll and register the file descriptors to it.
+  void init_epoll() {
+    epoll_fd = epoll_create1(0);
+    if (epoll_fd == -1)
+      error(errno, "error creating epoll");
+
+    auto add_fd = [&](fd_t fd) {
+      logmsg(LOG_DEBUG, "epoll will listen for fd %d", fd);
+      epoll_event ev{};
+      ev.data.fd = fd;
+      ev.events = EPOLLIN;
+      if (epoll_ctl(epoll_fd, EPOLL_CTL_ADD, fd, &ev))
+        error(errno, "failed to add fd %d to epoll", fd);
+    };
+
+    // Always listen for child exit events.
+    if (child_exited_pipe == -1)
+      error(0, "SIGCHLD handler not installed");
+    add_fd(child_exited_pipe);
+
+    // Listen for incoming data only when proxy is enabled.
+    if (has_proxy()) {
+      for (auto &proc : processes) {
+        add_fd(proc.process_to_proxy);
+      }
+    }
+  }
+
+  // Wait for an exited child without blocking. If a child has been waited
+  // successfully this method returns true.
+  bool handle_child_exit() {
+    // Consume what the signal handler wrote in this pipe.
+    static char buffer[1];
+    if (read(child_exited_pipe, buffer, 1) != 1) {
+      // This function may be called also if no one wrote in the pipe, so ignore
+      // those errors but still try to wait for a child.
+      if (errno != EAGAIN && errno != EWOULDBLOCK)
+        error(errno, "failed to read from exit pipe");
+    }
+
+    int status = -1;
+    // Check if a child exited without blocking.
+    pid_t pid = waitpid(-1, &status, WNOHANG);
+    if (pid < 0)
+      error(errno, "failed to wait for child exit");
+    // No child has exited.
+    if (pid == 0)
+      return false;
+
+    logmsg(LOG_DEBUG, "child with pid %d exited", pid);
+
+    if (first_process_exit_id == -1)
+      first_process_exit_id = pid;
+
+    // Search the exited process and store its exit information.
+    bool found = false;
+    for (auto &proc : processes) {
+      if (proc.pid != pid)
+        continue;
+
+      proc.on_exit(status);
+      found = true;
+      break;
+    }
+
+    // One of the process exited, close all the fd. `close_fds` must be called
+    // after `on_exit`.
+    for (auto &proc : processes) {
+      proc.close_fds();
+    }
+
+    if (!found)
+      error(0, "unknown child with pid %d exited", pid);
+
+    return true;
+  }
+
+  // Check if every process has exited.
+  bool has_everyone_exited() {
+    for (const auto &proc : processes)
+      if (!proc.exited)
+        return false;
+    return true;
+  }
+
+  // The pipe connecting from -> to has some data ready. Consume it reading as
+  // much as possible, copy it to the target process and write it to the output
+  // file.
+  void pump_proxy_pipe(const process_t &from, const process_t &to,
+                       output_file_t &output_file) {
+    const size_t BUF_SIZE = 1024 * 1024;
```

Comment:
Did you try with other values here? I think it doesn't matter too much but probably 4k is best since it's the page size.

Reply:
I haven't tried it actually. That value is the same as the previous implementation. 4k will be probably enough for most of the cases. Another option could be to keep the buffer very large, but allocate that buffer statically (this is a single-threaded program, so there shouldn't be any concurrency issues regarding that.)